packages/components/src/tree-grid/index.js

```diff
@@ -151,10 +154,10 @@ function TreeGrid(
 
 				// Focus the next column only if expanded.
 				if (
-					nextIndex !== 0 &&
+					keyCode === RIGHT &&
 					activeRow.getAttribute( 'aria-expanded' ) === 'false'
 				) {
-					nextIndex -= 1;
+					return;
 				}
```

Comment:
I think the problem I have reviewing here is that I'm not really sure what the desired behavior is.

Should the right key always be prevented from doing anything when a row is expanded?

Something that needs to be considered is that the `TreeGrid` component is general purpose. List View has two columns, but another implementation of TreeGrid might have more than two columns, so any code added should take that into account.

I could have a TreeGrid with four columns. Say that third column of a collapsed row is currently highlighted. Is the expectation that pressing right does nothing?

Reply:
@talldan The problem I am trying to solve is this.
1. I only want the Block selection button to be able to expand and collapse the child Blocks. In the List View, that is the 1st column with index of 0.
2. The current model is not very user friendly for screen reader users. To expand, you must find the collapsed Block, press Right Arrow to focus the Options button, and then Right Arrow again to expand. The problem is, all the way at the right, the user will not hear the expanded state get updated because Options button contains:

```
aria-expanded="false"
```

Because the Options menu for the Block is closed. The only way to know for sure that the row was expanded again is to either navigate with Down Arrow to the first child Block Options button or to press Left Arrow and you should hear the expanded state on the Block selection button.

To be honest, I think this needs a dedicated component as what we're trying to do here is really pretty complex. If not, I have another option that may work.
1. Expand the List View to 3 columns.
2. Block name column which will be used to expand/collapse, link that jumps to the Block, and finally the Block Options button.
3. This way we don't have to worry about relying on table position to expand/collapse. We simply listen for an onClick for the first column.

It may be easier to show you why this doesn't work in practice. If you would like, happy to jump on a quick Zoom and I can show you the big UX improvement between trunk and my branch. Maybe that would be better, that way you can visualize exactly what the problem is?

Thanks for hanging in there with me.

Reply:
I hadn't considered the issue with the options button having its own `aria-expanded`.

Testing again now and I think the only remaining issue I see in List View is that it's possible to this:
1. Add a paragraph and a columns block
2. Focus the columns block in List View and collapse it
3. Press the up key (focus the paragraph), then the right key (focus the options button of the paragraph), then the down key (focus the options button of the columns block)
4. Press the right key

Expected: Nothing happens
Actual: The collapsed column is expanded.

The issue here might be screen reader announcements, since the options button's `aria-expanded` won't change.

Maybe the right arrow key to expand should only work in the first column.

Reply:
Yes, that's what I am trying to say. To expand should only happen in the 1st column. The 1st column should be responsible for collapsing and expanding.

Reply:
I'm not offering much more than has been said, but I've been comparing this PR's changes with trunk.

On trunk the expand does indeed occur on the second td element with a right arrow click.

This PR changes the expand to the first td element.

I can reproduce Dan's scenario above, but it appears to be consistent for all options buttons, that is, I cannot right arrow (only down arrow to the lower blocks' options, or left to select the row.

This appears fairly intuitive to me as I'm at the end item of a single row. The caveat is that I mainly navigate using a pointer. :)

Reply:
@ramonjd Any ideas on how to keep this same functionality and maybe clean the code up a bit? Ultimately, the 1st column should do the expanding/collapsing. I am not worried about any interaction besides this as this will bring a greatly improved UX experience. Got to keep in mind about also making this not effect any uses of TreeGrid in the future. That has been my challenge with this PR.

Reply:
Hi @alexstine Thanks for thinking about this challenge. 

I've had a play around and I think I understand things a little better. I think the PR works the way we want to, but I do agree with the particular concern that other implementations of TreeGrid might have a column count greater than 2.

If I've understood, we want to tell the TreeGrid to expand at a specific column for right arrow clicks, rather than the default, which is the last column.

Given that, could we not pass TreeGrid an instruction saying that we want to trigger an expand at a specific column index?

The following code and naming is a bit rough, but the code is for illustration purposes. :)

### Example code, passing a prop to TreeGrid:

```jsx
<TreeGrid
    columnIndexToTriggerExpand={ 0 }
    ...otherProps,
>
```
### Example code in tree-grid/index.js:

```js
// Trigger an expand at a specific row if columnIndexToTriggerExpand
if (
    nextIndex !== currentColumnIndex &&
    activeRow.getAttribute( 'aria-expanded' ) === 'false' &&
    columnIndexToTriggerExpand === currentColumnIndex &&
    keyCode === RIGHT
   ) {
    onExpandRow( activeRow );
    event.preventDefault();
    return;
}

// Focus is either at the left or right edge of the grid. Allow Right Arrow to expand the row.
if ( nextIndex === currentColumnIndex ) {
    if ( keyCode === LEFT ) {
```

So when the prop `columnIndexToTriggerExpand` is not passed to TreeGrid the code wouldn't execute.

Reply:
@ramonjd The only problem with an approach like this is the way the current column index is working. If I am in the Options column and I press Left Arrow, it will return a current index of 1 which makes no sense because I should not be in the column anymore. None the less, it doesn't work for whatever reason. Calculating indexes like this is turning out to be really unreliable when you want a specific column to do the expanding. It will only return 0 once you are at the edge of the left column or something like that. I don't have vision so can't honestly tell you exactly what it is doing.

How about instead of passing a prop, I attach a data-column-trigger="true"? I know @talldan didn't want to go that way but for me, it seems like it would be safer. That way I can always know that data tag when focus is placed on it, the collapse/expand can work.

Thoughts?

Reply:
That's a fair point, thanks for explaining. I was, in a clumsy way, trying to advocate for some sort of flag to tell TreeGrid to expand at a certain point.

> How about instead of passing a prop, I attach a data-column-trigger="true"?

This sounds like it's worth a shot - I would tend to trust @talldan's judgement on this, but it would be good to see working. Maybe he'll come around ;)

Reply:
> The only problem with an approach like this is the way the current column index is working. If I am in the Options column and I press Left Arrow, it will return a current index of 1 which makes no sense because I should not be in the column anymore. None the less, it doesn't work for whatever reason.

This is the same function that handles the arrow key press and moves focus, so `currentColumnIndex` is always the index before focus has moved.`nextIndex` is where focus will end up if the function doesn't return early.

I think @ramonjd's code should work. I wonder if it's needed though, with what's being proposed it seems like column zero would always be the column where expands/collapse works.

The trouble is that this code previously just handled moving focus, but now more stuff has been added to it, and it's a little confusing. If this code were to be rewritten, I think it might be best to assign some well named variables to make the code more readable:
```jsx
const isCollapsed = activeRow.getAttribute( 'aria-expanded' ) === 'false';
const lastColumnIndex = focusablesInRow.length - 1;
const shouldExpand = isCollapsed && currentColumnIndex === 0 &&  keyCode === RIGHT;
const shouldCollapse = ! isCollapsed && currentColumnIndex === 0 && keyCode === LEFT;
const shouldMoveLeft = ! shouldCollapse && currentColumnIndex > 0 && keyCode === LEFT;
const shouldMoveRight = ! shouldExpand && currentColumnIndex < lastColumnIndex && keyCode === RIGHT;
```

Reply:
@talldan I thought at one point you were worried about having another column that could possibly be the one doing the expanding/collapsing. E.g.

```
<table>
<tr>
<td></td>
<td>Expand/Collapse</td>
</tr>
<tr></tr>
</table>
```

That is why I was working on extra code to try to add support for any column specified by a developer but adapt it for the List View. If you are okay with column index of 0 (first column) doing all the expanding and collapsing, let me clean the code up.

Reply:
Probably a miscommunication. I'm ok with it just being the first column 👍